src/OpenTelemetry/Metrics/Tags.cs

```diff
@@ -52,33 +34,100 @@ public Tags(KeyValuePair<string, object?>[] keyValuePairs)
 
     public readonly bool Equals(Tags other)
     {
-        var length = this.KeyValuePairs.Length;
+        var ourKvps = this.KeyValuePairs;
+        var theirKvps = other.KeyValuePairs;
 
-        if (length != other.KeyValuePairs.Length)
+        if (ourKvps.Length != theirKvps.Length)
         {
             return false;
         }
 
-        for (int i = 0; i < length; i++)
+#if NET6_0_OR_GREATER
+        // Note: This loop uses unsafe code (pointers) to elide bounds checks on
+        // two arrays we know to be of equal length.
+        var cursor = ourKvps.Length;
+        if (cursor > 0)
+        {
+            ref var ours = ref MemoryMarshal.GetArrayDataReference(ourKvps);
+            ref var theirs = ref MemoryMarshal.GetArrayDataReference(theirKvps);
+            while (true)
```

Comment:
Could you simplify this with a `for` loop?

```csharp
if (ourKvps.Length > 0)
{
    ref var ours = ref MemoryMarshal.GetArrayDataReference(ourKvps);
    ref var theirs = ref MemoryMarshal.GetArrayDataReference(theirKvps);

    for (int i = 0; i < ourKvps.Length; i++)
    {
        / Note: string.Equals performs an ordinal comparison
        if (!ours.Key.Equals(theirs.Key))
        {
            return false;
        }
        
        if (!ours.Value?.Equals(theirs.Value) ?? theirs.Value != null)
        {
            return false;
        }

        ours = ref Unsafe.Add(ref ours, 1);
        theirs = ref Unsafe.Add(ref theirs, 1);
    }
}
```

Reply:
It will be slightly slower to do that for loop style.

In the existing one we have this...

```csharp
while (true)
{
        ...

        if (--cursor == 0)
        {
             break; // Break out of loop here if done and skip the increment of pointers
        }

        ours = ref Unsafe.Add(ref ours, 1);
        theirs = ref Unsafe.Add(ref theirs, 1);
}
```

The for loop has...

```csharp
    for (int i = 0; i < ourKvps.Length; i++)
    {
        // ...

        ours = ref Unsafe.Add(ref ours, 1);
        theirs = ref Unsafe.Add(ref theirs, 1);
    }
```

It will be slightly slower because during the last iteration it will execute those two pointer increment ops for no reason.